Makefile

```diff
@@ -178,6 +193,24 @@ lib/libstdc++.a: $(shell $(CC) -print-file-name=libstdc++_pic.a)
 	@ar rcs $@ .libstdc++/*.o
 	@rm -r .libstdc++
 
+lib/libc++.a: $(shell $(CC) -print-file-name=libc++_p.a)
+	@echo "Frobnicating   $@"
+	@mkdir -p lib
+	@rm -rf .libc++
+	@mkdir .libc++
+	@(cd .libc++ && ar x $<)
+	@for i in .libc++/*.o ; do \
+		nm $$i | grep -q \@ || continue ; \
+		nm $$i | awk '$$3 ~ /@@/ { COPY = $$3; sub(/@@.*/, "", COPY); print $$3, COPY; }' > .libc++/replacements ; \
+		objcopy --redefine-syms=.libc++/replacements $$i $$i.new && mv $$i.new $$i ; \
+		rm .libc++/replacements ; \
+		nm $$i | awk '$$3 ~ /@/ { print $$3; }' > .libc++/deletes ; \
+		objcopy --strip-symbols=.libc++/deletes $$i $$i.new && mv $$i.new $$i ; \
+		rm .libc++/deletes ; \
+	done
+	@ar rcs $@ .libc++/*.o
+	@rm -r .libc++
```

Comment:
This whole section exists for libstdc++ to deal with the fact that we're trying to make it so that binaries built for the release can be run on very old linux systems, so we need to control the versions of symbols pulled from glibc.  libstdc++ will sometimes be compiled such that it requests certain versions of symbols, so we need to strip that out.

Since FreeBSD doesn't have glibc, this might not be an issue?  I have no idea if libc has versioned symbols or not, and if libc++ on FreeBSD would have the same versioned symbols problem or not.

Reply:
Yes, I think I've run into this before usually if I move binaries between major versions. Lately 10 has been the lowest version of FreeBSD that I see in heavier use and 12 will be a lot more common soon. I'll definitely look into this more there is some info on the mailing list discussing it a little here: https://lists.freebsd.org/pipermail/freebsd-ports/2016-August/104441.html

Reply:
As much as I would love to support this feature I think that it's going to add unnecessary overhead. So I found libc_pic.a finally and when trying to link against that I get:

    /usr/lib/libc_pic.a(arc4random.pico): In function `arc4random':
    /usr/src/lib/libc/gen/arc4random.c:235: multiple definition of `arc4random'
    /usr/local/lib/libcrypto.a(arc4random.o):(.text+0x0): first defined here
    /usr/lib/libc_pic.a(arc4random.pico): In function `arc4random_buf':
    /usr/src/lib/libc/gen/arc4random.c:247: multiple definition of `arc4random_buf'
    /usr/local/lib/libcrypto.a(arc4random.o):(.text+0x80): first defined here
    /usr/bin/ld: lib/libfdb_c.so: version node not found for symbol _malloc_options@FBSD_1.0

Well, FreeBSD 1.0 is version 7 which was before Clang became it's compiler by default. I'm going to possibly open this can of worms later and take your advice here and move on from that. Really, it would mean consistently writing a lot of FreeBSD specific code into the .map file. There is more on this subject here: https://wiki.freebsd.org/BinaryCompatibility

Reply:
@alexmiller-apple I've been trying to clean up everything and wrap this up but having some trouble with this part and the bindings. I'd just like to say that FDB has improved amazingly since I started this. The latest code just builds so much faster it's a lot easier to work on. So this section:

    lib/libc++.a: $(shell $(CC) -print-file-name=libc++_p.a)
	@echo "Frobnicating   $@"
	@mkdir -p lib
	@rm -rf .libc++
	@mkdir .libc++
	@(cd .libc++ && ar x $<)
	@for i in .libc++/*.o ; do \
		nm $$i | grep -q \@ || continue ; \
		nm $$i | awk '$$3 ~ /@@/ { COPY = $$3; sub(/@@.*/, "", COPY); print $$3, COPY; }' > .libc++/replacements ; \
		objcopy --redefine-syms=.libc++/replacements $$i $$i.new && mv $$i.new $$i ; \
		rm .libc++/replacements ; \
		nm $$i | awk '$$3 ~ /@/ { print $$3; }' > .libc++/deletes ; \
		objcopy --strip-symbols=.libc++/deletes $$i $$i.new && mv $$i.new $$i ; \
		rm .libc++/deletes ; \
	done
	@ar rcs $@ .libc++/*.o
	@rm -r .libc++

After I completely removed it I noticed the bindings are now non-functional so there is some issue with this branch here. I try to run some tests, start the server, and check the bindings but now I'm getting an error about undefined symbols. How much of what's going on here would affect the bindings if anything at all? I've narrowed it down to being this or something else but seems strange to me. The error message in the Python bindings is:

    OSError: /usr/home/jessebennett/fdb-branch/foundationdb/lib/libfdb_c.so: Undefined symbol "_ZTVN10__cxxabiv117__class_type_infoE 

I'll probably try some other things too later but everything builds as usual. This is probably the last thing I'm addressing before submitting for final review.

Reply:
@alexmiller-apple Ok, finally understand what you are saying here I had to look closer (scroll back farther). Unfortunately and fortunately, changes to this are not the cause of the problem with bindings right now but this has been noted:

    Frobnicating   lib/libc++.a
    nm: '.libc++/*.o': No such file or directory
    ar: warning: can't open file: .libc++/*.o: No such file or directory

Reply:
I would think that you don't need the libc++ "frobnication", because it's just trying to remove symbol versions from glibc, which you're not going to have on FreeBSD anyway.

Looking at the error that you got before, it looks like you're missing linking to libc++abi ?  If you just add `-lc++abi`, does that solve your problem?